h2o-core/src/test/java/water/fvec/VecTest.java

```diff
@@ -148,4 +146,50 @@ public void map(Chunk c) {
     source.remove(new Futures()).blockForPending();
     con.remove(new Futures()).blockForPending();
   }
+
+  @Test public void testChunkForChunkIdxAfterVecUpdate() {
+    try {
+      Scope.enter();
+      Vec v1 = Scope.track(makeSeq(2 * FileVec.DFLT_CHUNK_SIZE, false));
+      Vec v2 = Scope.track(makeSeq(2 * FileVec.DFLT_CHUNK_SIZE, false));
+      Chunk c0 = v1.chunkForChunkIdx(0);
+      c0._vec = v2; // inject any vec into the cached POJO, this simulates a stale Vec reference
+      Chunk c0c = v1.chunkForChunkIdx(0);
+      assertSame(c0, c0c);
+      // the vec reference was updated in the existing live object
+      assertSame(v1, c0c._vec);
+    } finally {
+      Scope.exit();
+    }
+  }
+
+  @Test public void testChunkForChunkIdxMRTask() {
+    Assume.assumeTrue(H2O.getCloudSize() > 0);
+    try {
+      Scope.enter();
+      Vec v = Scope.track(makeConN((long) 1e6, 16));
+      // 1. Run an arbitrary MRTask to populate the POJO Chunk caches
```

Comment:
@michalkurka What I'm trying to understand is what are you referring to by saying "POJO Chunk caches". Is it just Chunk type objects in the heaps of the Nodes where MRTask is being executed? And what you are saying is that by updating newly retrieved from DKV Vec's instances on the remote Node could be not the ones that will be used in second MRTask ( stage 3 in you test below)?

Reply:
@deil87 What you actually put in DKV is Value objects (think of them as wrappers), not the POJOs directly. Value object has a raw byte array (iced object) and a live object POJO, this is what I refer to as a cache. The live POJO of a Chunk can be discarded to release memory if the node is under memory pressure. New POJO will be re-instantiated from the raw bytes when needed again.

Reply:
@michalkurka 
> You have no guarantee that the Vec you are updating will actually be in the cache on the node in both Chunk and in the DKV cache.

So Chunk Value object stores live Chunk POJO in it and DKV stores Vec Value object with both byte array and live Vec POJO.  When we update Vec in DKV it will only update Vec Value object (and probably discard live Vec's POJO) but there is no logic that will go to Chunk Value object and update/discard live chunk's POJO which stores reference to old live Vec's POJO, right?

Reply:
> but there is no logic that will go to Chunk Value object and update/discard live chunk's POJO which stores reference to old live Vec's POJO, right?

Yes, that is absolutely correct.

When you update a Vec in DKV (by DKV.put for example) you will update the live object (POJO) on the node that made the update, you will update the bytes on both node who made the update and on the home node. On the home node, the POJO will be invalidated, on other nodes will just invalidate their entry in local H2O.STORE.

After the update, any node that runs DKV.getGet(vec key) will get a correct updated Vec. So far so good.

But none of this involves Chunks... we were only concerned with the content of H2O.STORE, we never touched any Chunks in the process. Which means the cached Vecs in the live chunk POJOs can have a stale instance of the Vec that is not actually reachable from anywhere else.

Because you can only get to a Chunk instance from `Vec#chunkForChunkIdx` this a good place to fix it. There are alternative solutions as well... but I find this approach reasonable and it has no performance harm.

Reply:
Cascades !

Reply:
Update Cascades ?! Leveraging the strategy pattern seems to be the natural choice during update operations ?